In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv("creditcard.csv")

In [4]:
data.shape

(284807, 31)

In [7]:
data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [11]:
data['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

In [12]:
## Dependent and Independent features

x = data.drop("Class", axis=1)
y = data['Class']

In [19]:
from sklearn.model_selection import train_test_split

In [20]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state=42)

#                                  ## Under Sampling ##

In the under-sampling we reduce the number of datapoints of class which has higher points 
EX for above dataset we have 280000 zero's and only 492 one's so we reduce zero's to 500 or something to make both classes equal with data points. But the disadvantage is that here we lost the data points that is why we usally do not use it.
We can use undersampling when our dataset is very very less
    
    

In [ ]:
from imblearn.under_sampling import NearMiss

ns = NearMiss(0.8) #Reduce the x_train untill the 80% of reduced data set = y_train (346/0.8 = 433)
X_train_us, Y_train_us = ns.fit_sample(x_train, y_train)

## Over Sampling
 In the over-sampling we increase the one's by given percentage parameter to make both classes compatable. 

In [ ]:
from imblearn.over_sampling import RandomOverSampler

os = RandomOverSampler(0.5) #increase the y_train untill the 50% of x_train (199017 * 0.5)
X_train_us, Y_train_us = os.fit_sample(x_train, y_train)

# # SMOTE Tomek

Do research
